linkerd/lock/src/shared.rs

```diff
@@ -76,15 +76,25 @@ impl<T> Shared<T> {
 
     pub fn release_and_notify(&mut self, value: T) {
         trace!(waiters = self.waiters.len(), "Releasing");
-        debug_assert!(match self.state {
+        assert!(match self.state {
             State::Claimed => true,
             _ => false,
         });
         self.state = State::Unclaimed(value);
         self.notify_next_waiter();
     }
 
-    pub fn notify_next_waiter(&mut self) {
+    pub fn release_waiter(&mut self, wait: Wait) {
+        // If a waiter is being released and it does not have a notify, then it must be being
+        // relased after being notified. Notify the next waiter to prevent deadlock.
```

Comment:
typo:
```suggestion
        // released after being notified. Notify the next waiter to prevent deadlock.
```
(are you having keyboard issues? :wink:)

Reply:
more like brainworms